homeassistant/components/light/tradfri.py

```diff
@@ -69,44 +98,70 @@ def brightness(self):
         """Return the brightness of the group lights."""
         return self._group.dimmer
 
-    def turn_off(self, **kwargs):
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs):
         """Instruct the group lights to turn off."""
-        return self._group.set_state(0)
+        result = yield from self._api(self._group.set_state(0))
+        return result
 
-    def turn_on(self, **kwargs):
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
         """Instruct the group lights to turn on, or dim."""
         if ATTR_BRIGHTNESS in kwargs:
-            self._group.set_dimmer(kwargs[ATTR_BRIGHTNESS])
+            yield from self._api(
+                self._group.set_dimmer(kwargs[ATTR_BRIGHTNESS]))
         else:
-            self._group.set_state(1)
+            yield from self._api(self._group.set_state(1))
 
-    def update(self):
-        """Fetch new state data for this group."""
-        self._group.update()
+    def _start_observe(self, err=None):
+        """Start observation of light."""
+        if err:
+            _LOGGER.info("Observation failed for {}".format(self._name), err)
 
+        observe_command = self._group.observe(callback=self._observe_update,
+                                              err_callback=self._start_observe,
+                                              duration=0)
+        observe_task = self._api(observe_command)
+        ensure_future(observe_task, loop=self._hass.loop)
 
-class Tradfri(Light):
+    def _refresh(self, group):
+        """Refresh the light data."""
+        self._group = group
+        self._name = group.name
+
+    def _observe_update(self, tradfri_device):
+        """Receive new state data for this light."""
+        self._refresh(tradfri_device)
+
+        self.schedule_update_ha_state()
+
+
+class TradfriLight(Light):
     """The platform class required by Home Asisstant."""
 
-    def __init__(self, light):
+    def __init__(self, light, api, hass):
         """Initialize a Light."""
-        self._light = light
-
-        # Caching of LightControl and light object
-        self._light_control = light.light_control
-        self._light_data = light.light_control.lights[0]
-        self._name = light.name
+        self._hass = hass
```

Comment:
`hass` will be set on the entity when added, so you don't need to pass `hass` in, you can just access it as an  instance attribute.